File: hashmap_linkedlist.py
'''

Your implementation should include these three functions: 

put(key, value) : Insert a (key, value) pair into the HashMap or If the value already exists, update the value.
get(key): Returns the value to which the specified key is mapped, or -1 if  no mapping for the key.
remove(key) : Remove the mapping for the value key if this map contains the mapping for the key
'''
# Time Complexity : Put() - O(n/m)
# Space Complexity : O(m+n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : Yes
'''
The put operation incorporates overwriting a key value pair if the same key is put again, The only way I could do that was
by traversing the linked list (corresponding to a particular hashed key). I could think of a better approach than that

Also, Selecting m was also quetionable, I tried Chain Doubling but that kept giving eoor in Leetcode, For now, I chose m = 1000, as hinted on leetcode question
'''

class Node():

    def __init__(self, key,value, next_=None):

        self.key = key
        self.value = value
        self.next = next_

class LinkedList():

    def __init__(self):
        self.head = None

    def add(self, key, value):

        node_data = Node(key, value)
        ll_iterator = self.head
        if self.head is None:
            self.head = node_data
        else:
            while ll_iterator:
                if key == ll_iterator.key:
                    ll_iterator.key = key
                    ll_iterator.value = value
                    break
                    
                prev = ll_iterator
                ll_iterator = ll_iterator.next
            else:
                prev.next = node_data
            


    def find_value(self, key):

        ll_iterator = self.head

        while ll_iterator:
            if ll_iterator.key == key:
                return ll_iterator.value

            ll_iterator = ll_iterator.next
            
        else:
            return -1
            
    def remove(self, key):

        ll_iterator = self.head

        if ll_iterator.key == key:
            self.head = ll_iterator.next

        else:
            while ll_iterator.next and ll_iterator.key != key:
                prev = ll_iterator
                ll_iterator = ll_iterator.next
            
            prev.next = ll_iterator.next



class hashmap():

    m = 1000
    def __init__(self):

        self.hashlist = [None] * self.m
        
    def get_hash(self, key):

        hashed_key= (key % self.m)-1
        return hashed_key


    def put(self, key, value):

        hashed_key = self.get_hash(key)
        
        if self.hashlist[hashed_key] is None:
            linked_list_obj = LinkedList()
            linked_list_obj.add(key, value)
            self.hashlist[hashed_key] = linked_list_obj
        else:
            self.hashlist[hashed_key].add(key, value)


    def get(self, key):

        hashed_key = self.get_hash(key)
        linked_list_cell = self.hashlist[hashed_key]
        
        if linked_list_cell:
            return linked_list_cell.find_value(key)
        else:
            return -1
    
    def remove(self, key):
        
        hashed_key = self.get_hash(key)
        linked_list_cell = self.hashlist[hashed_key]
        linked_list_cell.remove(key)
        

hashMap = hashmap()
hashMap.put(1, 2)

hashMap.put(2, 3)

print (hashMap.get(1))   #       // returns 2
print (hashMap.get(3))     #       // returns -1 (not found)

hashMap.put(2, 4)   #       // update the existing value

print (hashMap.get(2))       #     // returns 4 

hashMap.remove(2)   #     // remove the mapping for 2

print (hashMap.get(2))#           // returns -1 (not found)

File: HashMap-LinkedList.java
//Exercise_1 : Design Hashmap--Using Linked List which is Linear Chaining
// Time Complexity : O(K)  -> for all operations where K will be size of linked list
// Space Complexity : O(N+K)=>Total Elements where N is size of parent array and K is the size of child list where value is stored(Linked List)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :  No


// Your code here along with comments explaining your approach

//Note: Here Single Hashing is implemented. I created a class Node and initialised array of Nodes. For each index I'm storing linked list of nodes  

// While traversing the linked list always return the prev node because, then operations such as get, put and remove can be done easily.



class MyHashMapLinkedList {

    /** Initialize your data structure here. */
    
    class Node{
        int key,value;
        Node next;
        
        Node(int key,int value){
            this.key = key;
            this.value = value;
        }
    }
    Node[] nodes;

    public MyHashMapLinkedList() {
        nodes = new Node[10000];//because child nodes length should be 100, if we will make parent to 10000 than child list will also be of 1000 and traversing 1000 nodes is not a good option
    }
    
    private int getIndex(int key){
        return Integer.hashCode(key)%nodes.length;//for finding th indexlest than 10000
    }
    
    private Node findPrev(Node head,int key){//because it will be common for get,put and remove. Once we get the prev node we can do anything
        Node curr = head;
        Node prev = null;
        
        while(curr!=null && curr.key!=key){
            
            prev = curr;
            curr = curr.next;
        }
        
        return prev;
    }
    
    /** value will always be non-negative. */
    public void put(int key, int value) {
        int index = getIndex(key);
        
        if(nodes[index]==null){
            nodes[index] = new Node(-1,-1);
        }
        
        Node prev = findPrev(nodes[index],key);
        
        if(prev.next==null){
            prev.next = new Node(key,value);
        }else{
            prev.next.value = value;
        }
        
    }
    
    /** Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key */
    public int get(int key) {
        int index = getIndex(key);
        
        if(nodes[index]==null){
            return -1;
        }
        
        Node prev = findPrev(nodes[index],key);
        
        if(prev.next==null){
            return -1;
        }
        
        return prev.next.value;
        
    }
    
    /** Removes the mapping of the specified value key if this map contains a mapping for the key */
    public void remove(int key) {
        
        int index = getIndex(key);
        
        if(nodes[index]==null){
            return ;
        }
        
        Node prev = findPrev(nodes[index],key);
        
        if(prev.next==null){
            return;
        }
        
        prev.next = prev.next.next;
    }
}
